File: mobile/kmp/ui/src/commonMain/kotlin/com/atls/hyperion/ui/theme/tokens/elevation/Elevations.kt
package com.atls.hyperion.ui.theme.tokens.elevation

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import com.atls.hyperion.ui.theme.tokens.colors.Colors

class Elevations(colors: Colors) {
    val xs: PressableElevationStates = Xs(colors.elevation.xs)
    val sm: PressableElevationStates = Sm(colors.elevation.sm)
    val md: PressableElevationStates = Md(colors.elevation.md)
    val lg: PressableElevationStates = Lg(colors.elevation.lg)
    val modal: ElevationStates = Modal(colors.elevation.md)

    private class Xs(color: Color) : PressableElevationStates {
        override val default = elevation(
            drop(color, offsetY = ShadowOffset.xs, blur = ShadowBlur.xs4)
        )
        override val pressed = elevation(
            inner(color, offsetY = ShadowOffset.xs, blur = ShadowBlur.xs5),
            drop(
                color,
                offsetY = ShadowOffset.xs,
                blur = ShadowBlur.xs5,
                spread = ShadowSpread.negativeXs
            )
        )
        override val disabled = elevation(
            drop(color, offsetY = ShadowOffset.xs, blur = ShadowBlur.xs5)
        )
    }

    private class Sm(color: Color) : PressableElevationStates {
        override val default = elevation(
            drop(color, offsetY = ShadowOffset.sm, blur = ShadowBlur.sm)
        )
        override val pressed = elevation(
            inner(color, offsetY = ShadowOffset.xs, blur = ShadowBlur.xs4),
            drop(
                color,
                offsetY = ShadowOffset.xs,
                blur = ShadowBlur.xs3,
                spread = ShadowSpread.negativeSm
            )
        )
        override val disabled = elevation(
            drop(color, offsetY = ShadowOffset.xs, blur = ShadowBlur.xs4)
        )
    }

    private class Md(color: Color) : PressableElevationStates {
        override val default = elevation(
            drop(color, offsetY = ShadowOffset.lg, blur = ShadowBlur.lg)
        )
        override val pressed = elevation(
            inner(color, offsetY = ShadowOffset.xs, blur = ShadowBlur.xs3),
            drop(
                color,
                offsetY = ShadowOffset.sm,
                blur = ShadowBlur.sm,
                spread = ShadowSpread.negativeXs
            )
        )
        override val disabled = elevation(
            drop(color, offsetY = ShadowOffset.sm, blur = ShadowBlur.xs)
        )
    }

    private class Lg(color: Color) : PressableElevationStates {
        override val default = elevation(
            drop(
                color,
                offsetY = ShadowOffset.xl,
                blur = ShadowBlur.lg,
                spread = ShadowSpread.negativeLg
            )
        )
        override val pressed = elevation(
            inner(color, offsetY = ShadowOffset.sm, blur = ShadowBlur.xs2),
            drop(
                color,
                offsetY = ShadowOffset.md,
                blur = ShadowBlur.md,
                spread = ShadowSpread.negativeLg
            )
        )
        override val disabled = elevation(
            drop(
                color,
                offsetY = ShadowOffset.md,
                blur = ShadowBlur.sm,
                spread = ShadowSpread.negativeMd
            )
        )
    }

    private class Modal(color: Color) : ElevationStates {
        override val default = elevation(
            drop(
                color,
                offsetY = ShadowOffset.xl3,
                blur = ShadowBlur.xl3,
                spread = ShadowSpread.negativeSm
            )
        )
    }
}

private fun elevation(vararg shadows: Shadow) = Elevation(shadows.toList())

private fun drop(
    color: Color,
    offsetX: Dp = ShadowOffset.none,
    offsetY: Dp,
    blur: Dp,
    spread: Dp = ShadowSpread.none
) = Shadow(
    type = ShadowType.Drop,
    offsetX = offsetX,
    offsetY = offsetY,
    blur = blur,
    spread = spread,
    color = color
)

private fun inner(
    color: Color,
    offsetY: Dp,
    blur: Dp
) = Shadow(
    type = ShadowType.Inner,
    offsetX = ShadowOffset.none,
    offsetY = offsetY,
    blur = blur,
    spread = ShadowSpread.none,
    color = color
)

File: ui/theme/src/tokens/elevation/elevations.ts
import type { Colors }           from '../colors/index.js'
import type { ElevationsTokens } from './interfaces.js'

import { shadowBlurs }           from './blurs.js'
import { shadowOffsets }         from './offsets.js'
import { shadowSpreads }         from './spreads.js'
import { createElevation }       from './utils/index.js'

export const elevations = (colors: Colors): ElevationsTokens => ({
  xs: {
    default: createElevation({
      color: colors.elevation.xs,
      offsetY: shadowOffsets.xs,
      blur: shadowBlurs.xs4,
    }),
    hover: createElevation(
      {
        color: colors.elevation.xs,
        offsetY: shadowOffsets.xs,
        blur: shadowBlurs.xs3,
        spread: shadowSpreads.xs,
      },
      {
        color: colors.action.subtle,
        offsetX: shadowOffsets.xs,
        offsetY: shadowOffsets.md,
        blur: shadowBlurs.sm,
        spread: shadowSpreads.negativeSm,
      }
    ),
    pressed: createElevation(
      {
        color: colors.elevation.xs,
        offsetY: shadowOffsets.xs,
        blur: shadowBlurs.xs5,
        inset: true,
      },
      {
        color: colors.elevation.xs,
        offsetY: shadowOffsets.xs,
        blur: shadowBlurs.xs5,
        spread: shadowSpreads.negativeXs,
      }
    ),
    disabled: createElevation({
      color: colors.elevation.xs,
      offsetY: shadowOffsets.xs,
      blur: shadowBlurs.xs5,
    }),
  },
  sm: {
    default: createElevation({
      color: colors.elevation.sm,
      offsetY: shadowOffsets.sm,
      blur: shadowBlurs.sm,
    }),
    hover: createElevation(
      {
        color: colors.elevation.sm,
        offsetY: shadowOffsets.sm,
        blur: shadowBlurs.md,
        spread: shadowSpreads.xs,
      },
      {
        color: colors.action.subtle,
        offsetX: shadowOffsets.xs,
        offsetY: shadowOffsets.lg,
        blur: shadowBlurs.md,
        spread: shadowSpreads.negativeSm,
      }
    ),
    pressed: createElevation(
      {
        color: colors.elevation.sm,
        offsetY: shadowOffsets.xs,
        blur: shadowBlurs.xs4,
        inset: true,
      },
      {
        color: colors.elevation.sm,
        offsetY: shadowOffsets.xs,
        blur: shadowBlurs.xs3,
        spread: shadowSpreads.negativeSm,
      }
    ),
    disabled: createElevation({
      color: colors.elevation.sm,
      offsetY: shadowOffsets.xs,
      blur: shadowBlurs.xs4,
    }),
  },
  md: {
    default: createElevation({
      color: colors.elevation.md,
      offsetY: shadowOffsets.lg,
      blur: shadowBlurs.lg,
    }),
    hover: createElevation(
      {
        color: colors.elevation.md,
        offsetY: shadowOffsets.lg,
        blur: shadowBlurs.xl,
      },
      {
        color: colors.action.subtle,
        offsetX: shadowOffsets.xs,
        offsetY: shadowOffsets.xl,
        blur: shadowBlurs.xl,
        spread: shadowSpreads.negativeLg,
      }
    ),
    pressed: createElevation(
      {
        color: colors.elevation.md,
        offsetY: shadowOffsets.xs,
        blur: shadowBlurs.xs3,
        inset: true,
      },
      {
        color: colors.elevation.md,
        offsetY: shadowOffsets.sm,
        blur: shadowBlurs.sm,
        spread: shadowSpreads.negativeXs,
      }
    ),
    disabled: createElevation({
      color: colors.elevation.md,
      offsetY: shadowOffsets.sm,
      blur: shadowBlurs.xs,
    }),
  },
  lg: {
    default: createElevation({
      color: colors.elevation.lg,
      offsetY: shadowOffsets.xl,
      blur: shadowBlurs.lg,
      spread: shadowSpreads.negativeLg,
    }),
    hover: createElevation(
      {
        color: colors.elevation.lg,
        offsetY: shadowOffsets.xl2,
        blur: shadowBlurs.xl2,
        spread: shadowSpreads.negativeLg,
      },
      {
        color: colors.action.subtle,
        offsetX: shadowOffsets.xs,
        offsetY: shadowOffsets.xl,
        blur: shadowBlurs.md,
        spread: shadowSpreads.negativeXs,
      }
    ),
    pressed: createElevation(
      {
        color: colors.elevation.lg,
        offsetY: shadowOffsets.sm,
        blur: shadowBlurs.xs2,
        inset: true,
      },
      {
        color: colors.elevation.lg,
        offsetY: shadowOffsets.md,
        blur: shadowBlurs.md,
        spread: shadowSpreads.negativeLg,
      }
    ),
    disabled: createElevation({
      color: colors.elevation.lg,
      offsetY: shadowOffsets.md,
      blur: shadowBlurs.sm,
      spread: shadowSpreads.negativeMd,
    }),
  },
  modal: {
    default: createElevation({
      color: colors.elevation.md,
      offsetY: shadowOffsets.xl3,
      blur: shadowBlurs.xl3,
      spread: shadowSpreads.negativeSm,
    }),
  },
})
